### ISSUE TITLE
Not Able To Change Status Of Invoice

### ISSUE THREAD

**Issue by amitpanwar86** [paypal/PayPal-PHP-SDK#328]
HI,

I am facing a problem in marking invoice as `MARKED_AS_PAID`.
Here is my code :

````php
$invoice=new Invoice();
$invoice->setId($invoiceId);
$invoice->setStatus("MARKED_AS_PAID");
$updateStatus=$invoice->update($apiContext);
return true;
````

But i am getting an error Got Http response code 400 when accessing https://api.sandbox.paypal.com/v1/invoicing/invoices/*************. I am stuck in this issue.
Can anyone please tell whats wrong with this???

Many Thanks in Advance,

**Comment by jziaja** [comment]
Hi @amitpanwar86, the body of the 400 response should detail what information was missing that the API needed, which in this case could be the `merchant_info` details (which is also marked as **Required** in the [public docs](https://developer.paypal.com/docs/api/#update-an-invoice)).  However, when I tested providing that along with setting the status, I got back a 500 error with no details.  I'll report that bug to the Invoicing API team.

It looks like updating an invoice is best done if you first call `get()` and get the full invoice resource object, make your changes, and then call `update()` with the full, modified invoice object.  Can you give that a try and see if it resolves your issue?

**Comment by jaypatel512** [comment]
Thanks @jziaja ! 

The best way to catch those 400 exception details, that include pretty much everything you need to know about the error, please follow instructions here at wiki for [ Exception Handling ](https://github.com/paypal/PayPal-PHP-SDK/wiki/exception-%27PayPal%5CException%5CPayPalConnectionException%27-with-message-%27Got-Http-response-code-400-when-accessing)

**Comment by amitpanwar86** [comment]
I know exception handling..i am getting this error in exception. But if i want to edit an invoice where to set invoice id ???  i set invoice id using setId() function. is this the correct way???

**Comment by jaypatel512** [comment]
Hey @amitpanwar86 !

I was able to reproduce your issue. I have sent an email to invoice team asking about that. I will let you know as soon as I get an answer from them.

**Comment by amitpanwar86** [comment]
Thanks @jaypatel512 ,

I am also getting an error while deleting invoice :
My Code is: 
      $invoice=new Invoice();
      $invoice=$invoice->get($invoiceId, $this->_apiContext);
      $invoice->delete($this->_apiContext);
and Error i got is : 
Got Http response code 405 when accessing https://api.sandbox.paypal.com/v1/invoicing/invoices/**********

This error is only with delete function. because get() method giving me complete invoice object when i call delete() method error appears.

AND @jziaja : Same error with update as you told me earlier.


Many Thanks,

**Comment by amitpanwar86** [comment]
Hey @jaypatel512 ,

Mate, Any luck? any reply from invoice team??

**Comment by jaypatel512** [comment]
Actually yes. We are debugging that right now. I will update as soon as we have a solution.

**Comment by amitpanwar86** [comment]
HI @jaypatel512 ,

Mate, Any update?

**Comment by jaypatel512** [comment]
Hey @amitpanwar86 ! not yet unfortunately. I am still waiting for the internal team to reproduce the issue and debug the root cause.

**Comment by amitpanwar86** [comment]
Hey @jaypatel512. Below is more debugging info : 

This is how my invoice looks like when I get it from the REST API : 
{
  "id": "INV2-8Z2J-BNN3-9ERS-XVJF",
  "number": "0037",
  "status": "PAID",
  "merchant_info": {
    "email": "aboutmyspeech-facilitator@gmail.com",
    "first_name": "Developer",
    "last_name": "Test",
    "business_name": "A-1 Technology",
    "phone": {
      "country_code": "001",
      "national_number": "8288877119"
    },
    "address": {
      "line1": "Line 1",
      "city": "Portland",
      "state": "OR",
      "postal_code": "97217",
      "country_code": "US",
      "phone": {
        "country_code": "001",
        "national_number": "8288877119"
      }
    }
  },
  "billing_info": [
    {
      "email": "bst_28@yahoo.co.in",
      "business_name": "Aone",
      "address": {
        "line1": "H.no.616, IInd Floor",
        "line2": "H.no.616, IInd Floor",
        "city": "Mohali",
        "state": "PB",
        "postal_code": "160062",
        "country_code": "US"
      },
      "additional_info": "Pay Per Use Invoicessss"
    }
  ],
  "items": [
    {
      "name": "Home",
      "quantity": 10,
      "unit_price": {
        "currency": "USD",
        "value": "0.17"
      }
    },
    {
      "name": "Brain Games",
      "quantity": 10,
      "unit_price": {
        "currency": "USD",
        "value": "0.20"
      }
    }
  ],
  "invoice_date": "2015-06-22 PDT",
  "note": "Pay Per Use Invoice",
  "logo_url": "https:\/\/www.google.co.in\/images\/srpr\/logo11w.png",
  "total_amount": {
    "currency": "USD",
    "value": "3.70"
  },
  "metadata": {
    "created_date": "2015-06-22 21:36:08 PDT",
    "first_sent_date": "2015-06-22 21:36:10 PDT",
    "last_sent_date": "2015-06-22 21:36:10 PDT",
    "payer_view_url": "https:\/\/www.sandbox.paypal.com\/cgi_bin\/webscr?cmd=_pay-inv&viewtype=altview&id=INV2-8Z2J-BNN3-9ERS-XVJF"
  }
}

Below is the response I get from Paypal when running update : 

{
  "name": "VALIDATION_ERROR",
  "message": "Invalid request - see details.",
  "debug_id": "aa96c811161e2",
  "details": [
    {
      "field": "metadata.created_date",
      "issue": "Can not construct instance of java.util.Calendar from String value '2015-06-22 21:36:08 PDT': not a valid representation (error: Can not parse date \"2015-06-22 21:36:08 PDT\": not compatible with any of standard forms (\"yyyy-MM-dd'T'HH:mm:ss.SSSZ\", \"yyyy-MM-dd'T'HH:mm:ss.SSS'Z'\", \"EEE, dd MMM yyyy HH:mm:ss zzz\", \"yyyy-MM-dd\")) at line:1, column:1049"
    }
  ]
}

Hope this helps in debugging this issue!

**Comment by jaypatel512** [comment]
That helps greatly. Thanks.

**Comment by jaypatel512** [comment]
Hey @amitpanwar86 !

I was able to get some lead on this. Basically, the way to change the status of the invoice is to make a separate call as shown here: https://developer.paypal.com/webapps/developer/docs/api/#record-a-payment. You can use the SDK sample to easily do this: http://paypal.github.io/PayPal-PHP-SDK/sample/doc/invoice/RecordPayment.html

However, for now, we found another issue under it. There are few read-only fields that gets returned when doing a GET on invoice, that needs to be removed. we are working with internal team to see if we could get it fixed, however, that may take some time.

In the mean time, please do these few things to resolve this issue.

1. Once you get an invoice object, remove few of these fields.

```php
// Invoice object holds the $invoice GET
// Remove the Metadata, as it is only read only
$invoice->setMetadata(null);
// If you have discount on entire invoice, follow this. Else skip this if loop.
if ($invoice->getDiscount()) {
   // If you have set Discount during creation, remove Amount, and viceversa
    $invoice->getDiscount()->setAmount(null);
}

// For each Item we need to remove the tax "Amount" or "Percentage". Both cannot be present.
foreach($invoice->getItems() as $item) {
    $item->getTax()->setAmount(null);
    // Same thing for itemized discount
    if ($item->getDiscount()) {
        $invoice->getDiscount()->setAmount(null);
    }
}
```

So, basically, for any amount vs percentage, based on your settings when you created an invoice, you could remove one or the other. 

I will let you know once I hear more from the internal team.

**Comment by jaypatel512** [comment]
Hey @amitpanwar86 !

I think I was right on the previous fix. The Invoice team mentioned that both "percentage" and "value" amount for any object cannot be accepted. It should be either or. 

So, basically, if you run into any validation issues where it says, you need to set either value or percentage, you need to apply similar logic as shown above to remove one or the other. You could remove any object, or value by using `set<MethodName>(null);` as shown above. 

Let me know if this fixes your problem, if not, feel free to re-open this issue.

**Comment by amitpanwar86** [comment]
Hi @jaypatel512 ,

    // For each Item we need to remove the tax "Amount" or "Percentage". Both cannot be present.
    foreach($invoice->getItems() as $item) {
        $item->getTax()->setAmount(null);
        // Same thing for itemized discount
        if ($item->getDiscount()) {
            $item->getDiscount()->setAmount(null);
        }
    }

I think this is the way to set discount for already created invoice but i want to add discount to items
while creating a invoice.  Here is my code i have tried so far : 

This the invoice json:

     {"merchant_info":{"email":"########@gmail.com","first_name":"Developer","last_name":"Test","business_name":"########","phone":{"country_code":"001","national_number":"8288877119"},"address":{"line1":"Line 1","city":"Portland","state":"OR","postal_code":"97217","country_code":"US"}},"billing_info":[{"email":"#####@gmail.com","business_name":"AmySys","additional_info":"CIMBA Invoice","address":{"line1":"Amit Kumar","state":"newyork","postal_code":"12345","city":"newyork","country_code":"US","line2":"AAAA"}}],"note":"Test Invoice","items":[{"name":"Home","quantity":"10","discount":"70","unit_price":{"currency":"USD","value":"0.15"}},{"name":"Wellness","quantity":"10","discount":"70","unit_price":{"currency":"USD","value":"0.03"}}],"logo_url":"https://www.google.co.in/images/srpr/logo11w.png"}

 And This is the error i got from paypal : 

    [url:PayPal\Exception\PayPalConnectionException:private] =>  https://api.sandbox.paypal.com/v1/invoicing/invoices
    [data:PayPal\Exception\PayPalConnectionException:private] => 
    [message:protected] => Got Http response code 500 when accessing https://api.sandbox.paypal.com/v1/invoicing/invoices.
    [string:Exception:private] => 
    [code:protected] => 500

**Comment by jaypatel512** [comment]
As per the sample provided [here](http://paypal.github.io/PayPal-PHP-SDK/sample/doc/invoice/CreateInvoice.html), and the api specifications provided [here](https://developer.paypal.com/webapps/developer/docs/api/#invoiceitem-object), the discount is not a string, but a cost object. 

Something like this would work:
```php
$item1discount = new Cost();
$item1discount->setPercent("3");
$items[1]
    ->setName("Injection")
    ->setQuantity(5)
    ->setDiscount($item1discount)
    ->setUnitPrice(new Currency());
```
